### PR TITLE
Fix tf.signal.kaiser_window second derivative at beta == 0

### LINKED ISSUE #127245: tf.signal.kaiser_window loses the second derivative at beta zero
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13.5

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.signal.kaiser_window` returns an incorrect second derivative when `beta` is zero.

In the reproducer below, `beta = 3 + 2 * t`, and the input is evaluated at `t = -1.5`, so `beta` is exactly zero. The Kaiser window is defined using the modified Bessel function `I0`; since `I0` is an even entire function, the weighted Kaiser-window expression is smooth at `beta = 0`.

The forward value is computed correctly as `385.0`, but nested reverse-mode autodiff returns `0.0` for the second derivative with respect to `t`. The expected second derivative is approximately `-365.85185185185185`.

Expected behavior?

The second derivative of the weighted `tf.signal.kaiser_window` output at `t = -1.5` should be close to `-365.85185185185185`, not `0.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    y = tf.signal.kaiser_window(
        10,
        beta=3 + 2 * t,
        dtype=t.dtype,
    )
    weights = tf.constant(
        [1, 4, 9, 16, 25, 36, 49, 64, 81, 100],
        dtype=t.dtype,
    )
    return tf.reduce_sum(y * weights)

def reverse(fn):
    def grad(x):
        with tf.GradientTape() as tape:
            tape.watch(x)
            y = fn(x)
        return tape.gradient(y, x)
    return grad

x = tf.constant(-1.5, dtype=tf.float64)

forward = target(x)
actual = reverse(reverse(target))(x)
expected = -365.85185185185185

print("forward:", forward.numpy())
print("actual:", actual.numpy())
print("expected:", expected)

if abs(float(actual) - expected) > 1e-9:
    print("BUG REPRODUCED: tf.signal.kaiser_window loses the second derivative at beta zero")
else:
    print("not reproduced")
```

### Relevant log output

```shell
forward: 385.0
actual: 0.0
expected: -365.85185185185185
BUG REPRODUCED: tf.signal.kaiser_window loses the second derivative at beta zero
```

### PR DESCRIPTION
**Pull request by Phoebus-Liu**
`tf.signal.kaiser_window` returns a zero second derivative with respect to
`beta` at `beta == 0`, although the window is an even analytic function of
`beta` and therefore twice differentiable there. Reported in #127245.

The window is `I0(beta * s) / I0(beta)`, where `s` does not depend on `beta`
and `I0` is an even entire function, so its second derivative at `beta == 0`
is `(s^2 - 1) / 2` — nonzero. The implementation passed `abs(beta)` to
`bessel_i0e`. Because the gradient of `Abs` is `sign(x)` and the gradient of
`Sign` is identically zero, the second derivative of `abs(beta)` vanishes at
the origin, and that cusp propagates to the whole expression: the second
derivative of the window evaluated to exactly `0.0`.

This is not a defect in `bessel_i0e`; its own second derivative at `x == 0`
is `0.5`, which is correct.

`bessel_i0e` is an even function, so `beta` is now passed to it with its
original sign, which removes the cusp. The exponential scaling factor keeps
using `abs(beta)`, so negative `beta` values still produce the same result as
positive ones and the computation stays numerically stable for large `beta`.

With the reproducer from the issue (`beta = 3 + 2t`, `t = -1.5`) the second
derivative is now `-365.85185185185185`, matching the expected value from the
issue; previously it was `0.0`.

Backward compatibility was checked for `beta >= 0`. The forward values change
slightly, because the exponential scaling factor is now computed as
`beta_abs * (shape - 1)` rather than `beta_abs * shape - beta_abs`; the two
forms are algebraically identical, but the latter loses precision to
catastrophic cancellation near the center of the window when `beta` is large.
The largest relative differences against the current implementation are
~3.5e-15 in float64, ~2.0e-6 in float32 and ~1.6e-2 in float16 (both at
`beta = 60`), and the new values are the more accurate ones: measured against
a higher-precision reference the new form reduces the error by up to ~400x in
float64 for large `beta`.

For negative `beta` the result is now exactly equal to the result for
`|beta|`; before this change the values differed by a factor of
`exp(2 * |beta| * (1 - s))`.

Adds `test_kaiser_window_second_derivative_at_zero_beta`. The existing tests
only cover the forward values for positive and negative `beta`, not any
derivative, which is why this went unnoticed.

Fixes #127245

### CODE REVIEW

File: tensorflow/python/ops/signal/window_ops.py [hunk 1]
```diff
@@ -79,15 +79,21 @@ def kaiser_window(window_length, beta=12., dtype=dtypes.float32, name=None):
...
+    # calling `bessel_i0e` would introduce a cusp at beta == 0 (the second
+    # derivative of `abs` vanishes at the origin), which makes the second
+    # derivative of the window with respect to `beta` evaluate to zero.  The
+    # exponential factor still uses `abs(beta)`, so that negative `beta`
+    # values produce the same result as positive ones.
+    beta = math_ops.cast(beta, dtype=dtype)
+    beta_abs = math_ops.abs(beta)
+    num = beta * shape
+    window = math_ops.exp(beta_abs * shape - beta_abs) * (
         special_math_ops.bessel_i0e(num) / special_math_ops.bessel_i0e(beta))
```
[the remark this answers is not in the document]
Comment (Phoebus-Liu) [reply tensorflow/python/ops/signal/window_ops.py]: Good catch — fixed in a0e7c96a, the factor is now `beta_abs * (shape - one)`.

I measured both forms against a long double reference before making the change, and the factored form is indeed the more accurate one: at `beta = 1e4` in float64 the maximum relative error of the exponential factor drops from 7.4e-13 to 1.7e-15.

The tradeoff is that forward values are no longer bit-for-bit identical to the previous implementation. The largest relative differences are ~3.5e-15 in float64, ~2.0e-6 in float32 and ~1.6e-2 in float16 (at `beta = 60`); in each case the new value is closer to the exact result, so this is a precision improvement rather than a regression. The second derivative at `beta == 0` and the handling of negative `beta` are unaffected. The PR description has been updated accordingly.

**Review by dmiltr3 — APPROVED** [review]

**Comment by aydenhailey78-hub** [comment]
This offer looks interesting; could you share related opportunities?

**Review by dmiltr3 — APPROVED** [review]
The code changes look good and internal presubmit test suites have successfully passed. Approving.

**Comment by Phoebus-Liu** [comment]
Thank you, @dmiltr3, for the careful review and for approving this PR. I appreciate you running the internal presubmit suites to confirm the changes. I'll keep an eye out for the merge.
